ext/node/analyze.rs

```diff
@@ -111,67 +113,88 @@ impl<TCjsCodeAnalyzer: CjsCodeAnalyzer> NodeCodeTranslator<TCjsCodeAnalyzer> {
+    handled_reexports.insert(entry_specifier.clone());
+    let (reexports_to_handle_tx, mut reexports_to_handle_rx) =
+      deno_core::unsync::mpsc::unbounded_channel();
+    reexports_to_handle_tx
+      .send((entry_specifier.clone(), analysis.reexports))
+      .unwrap();
+    let mut analyze_futures: FuturesUnordered<
+      LocalBoxFuture<Result<_, AnyError>>,
+    > = FuturesUnordered::new();
+    while !reexports_to_handle_rx.is_empty() || !analyze_futures.is_empty() {
+      tokio::select! {
+        message = reexports_to_handle_rx.recv() => {
+          // 1. Resolve the re-exports and start a future to analyze it
+          let (referrer, reexports) = message.unwrap();
+          for reexport in reexports {
+            let reexport_specifier = self.resolve(
+              &reexport,
+              &referrer,
+              // FIXME(bartlomieju): check if these conditions are okay, probably
+              // should be `deno-require`, because `deno` is already used in `esm_resolver.rs`
+              &["deno", "require", "default"],
+              NodeResolutionMode::Execution,
+              permissions,
+            )?;
+            if !handled_reexports.insert(reexport_specifier.clone()) {
+              continue;
+            }
 
-      // Second, resolve its exports and re-exports
-      let analysis = self
-        .cjs_code_analyzer
-        .analyze_cjs(&reexport_specifier, None)
-        .await
-        .with_context(|| {
-          format!(
-            "Could not load '{}' ({}) referenced from {}",
-            reexport, reexport_specifier, referrer
-          )
-        })?;
-      let analysis = match analysis {
-        CjsAnalysis::Esm(_) => {
-          // todo(dsherret): support this once supporting requiring ES modules
-          return Err(anyhow::anyhow!(
-            "Cannot require ES module '{}' from '{}'",
-            reexport_specifier,
-            specifier
-          ));
+            let cjs_code_analyzer = &self.cjs_code_analyzer;
+            let referrer = referrer.clone();
+            analyze_futures.push(
+              async move {
+                let analysis = cjs_code_analyzer
+                  .analyze_cjs(&reexport_specifier, None)
+                  .await
+                  .with_context(|| {
+                    format!(
+                      "Could not load '{}' ({}) referenced from {}",
+                      reexport, reexport_specifier, referrer
+                    )
+                  })?;
+
+                Ok((reexport_specifier, referrer, analysis))
+              }
+              .boxed_local(),
+            );
+          }
         }
-        CjsAnalysis::Cjs(analysis) => analysis,
-      };
+        analysis_result = analyze_futures.select_next_some() => {
+          // 2. Look at the analysis result and resolve its exports and re-exports
+          let (reexport_specifier, referrer, analysis) = analysis_result?;
```

Comment:
Something I just realized is we should make the errors deterministic if there are multiple.

Reply:
Done in a1291ce6f186fec3bac2c25dbe1e6d7428bc3f4e